source/common/stats/heap_stat_data.cc

```diff
@@ -6,7 +6,10 @@
 namespace Envoy {
 namespace Stats {
 
-HeapStatData::HeapStatData(absl::string_view key) : name_(key.data(), key.size()) {}
+HeapStatData::HeapStatData(absl::string_view key) {
+  memcpy(name_, key.data(), key.size());
+  name_[key.size()] = '\0';
+}
```

Comment:
`snprintf(name_, key.size + 1, key.data); `  this is better?

> snprintf ... Writes the results to a character string buffer. (...) will be terminated with a null character, unless buf_size is zero.

Reply:
I'd prefer not to use a dynamic formatting function if I'm just copying bytes. I think what I have is OK but I'm an old C programmer. I'm happy to use a simpler API if there's a way to do this in one call without formatting.

Reply:
Actually it looks like this is what StringUtil::strlcpy is for, which essentially does what I had inlined here. Might as well use it.

Reply:
I'm curious for you an initialization char [] to memcpy will not have a problem?

Reply:
Asan error in the ci should be related to this

Reply:
I think you should use the RawSlice in  envoy replace char[]

Reply:
This is a way of allocating a variable-length structure, which must have some fixed- size elements, and the last element declared with [] is a variable-length array. You can't allocate it with a simple 'new' but you can manually allocate the memory (e.g with malloc) and used placed new.

Reply:
variable-length structure should `char[0]` or `char[1]` not char[]

Reply:
Is described in c++ specification does not support a zero-length of the array, in C99 only support.

When I test machine is the core

```
tianqian-zyf@ubuntu:~$ cat 6.cc 
#include <iostream>
#include <string.h>

struct test {
  test() {
      memcpy(data, "1234", 4);
      data[4] = '\0';
  }
char data[];
};


int main() {
  test t;
  std::cout << t.data << "\n";
}
tianqian-zyf@ubuntu:~$ g++-5 6.cc 
tianqian-zyf@ubuntu:~$ ./a.out 
1234
*** stack smashing detected ***: ./a.out terminated
Aborted (core dumped)
```

https://stackoverflow.com/questions/295027/array-of-zero-length
https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html

Reply:
I think the sanitization failure was due to my failing to correct the operator[] usage in `ThreadLocalStoreImpl::ScopeImpl::histogram()` -- I had done it correctly in counter() and gauge().  Good thing we have these sanitizers. Will fix.

I'll look up the official reference for the usage of `[]` but there's an informal reference here: https://www.geeksforgeeks.org/flexible-array-members-structure-c/ and there was this pattern of usage in Envoy for RawStatData before I started working on it.

Your example doesn't work because you have to manually size 'test', in your case you'd need to use `void* memory = malloc(sizeof(test) + 5)` to accommodate your usage.

Reply:
You are right, it is the way I use wrong, best can find the official reference for the usage of `[]` in c++ standard

Reply:
Here we go with an official-looking site: https://en.cppreference.com/w/c/language/array
```
Within a struct definition, an array of unknown size may appear as the last member (as long as there is at least one other named member), in which case it is a special case known as flexible array member. See struct for details:

struct s { int n; double d[]; }; // s.d is a flexible array member 
struct s *s1 = malloc(sizeof (struct s) + (sizeof (double) * 8)); // as if d was double d[8]
```